<a href="https://colab.research.google.com/github/Rlours/Airbnb-Listings/blob/main/Urban_Air_Pollution_00.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# First Basic Version

In [1]:
# Mounting Drive
from google.colab import drive

drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [2]:
# Importing Libraries
import pandas as pd

# Visualization libraries

import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
# Importing data
df_train = pd.read_csv('/content/drive/My Drive/Colab Notebooks/DF_Zindi_Urban-Air/Train.csv')
df_test = pd.read_csv('/content/drive/My Drive/Colab Notebooks/DF_Zindi_Urban-Air/Test.csv')
df_sub = pd.read_csv('/content/drive/My Drive/Colab Notebooks/DF_Zindi_Urban-Air/SampleSubmission.csv')

display(df_train.head())
print("Shape-Training Dataset")
print("df_train",df_train.shape)

print("Data Types :")
df_train.info(5)

Place_ID X Date        Date Place_ID  target  target_min  target_max  \
0  010Q650 X 2020-01-02  2020-01-02  010Q650    38.0        23.0        53.0   
1  010Q650 X 2020-01-03  2020-01-03  010Q650    39.0        25.0        63.0   
2  010Q650 X 2020-01-04  2020-01-04  010Q650    24.0         8.0        56.0   
3  010Q650 X 2020-01-05  2020-01-05  010Q650    49.0        10.0        55.0   
4  010Q650 X 2020-01-06  2020-01-06  010Q650    21.0         9.0        52.0   

   target_variance  target_count  precipitable_water_entire_atmosphere  \
0           769.50            92                             11.000000   
1          1319.85            91                             14.600000   
2          1181.96            96                             16.400000   
3          1113.67            96                              6.911948   
4          1164.82            95                             13.900001   

   relative_humidity_2m_above_ground  ...  L3_SO2_sensor_zenith_angle  \
0                          60.200001  ...                   38.593017   
1                          48.799999  ...                   59.624912   
2                          33.400002  ...                   49.839714   
3                          21.300001  ...                   29.181258   
4                          44.700001  ...                    0.797294   

   L3_SO2_solar_azimuth_angle  L3_SO2_solar_zenith_angle  \
0                  -61.752587                  22.363665   
1                  -67.693509                  28.614804   
2                  -78.342701                  34.296977   
3                  -73.896588                  30.545446   
4                  -68.612480                  26.899694   

   L3_CH4_CH4_column_volume_mixing_ratio_dry_air  L3_CH4_aerosol_height  \
0                                    1793.793579            3227.855469   
1                                    1789.960449            3384.226562   
2                                            NaN                    NaN   
3                                            NaN                    NaN   
4                                            NaN                    NaN   

   L3_CH4_aerosol_optical_depth  L3_CH4_sensor_azimuth_angle  \
0                      0.010579                    74.481049   
1                      0.015104                    75.630043   
2                           NaN                          NaN   
3                           NaN                          NaN   
4                           NaN                          NaN   

   L3_CH4_sensor_zenith_angle  L3_CH4_solar_azimuth_angle  \
0                   37.501499                  -62.142639   
1                   55.657486                  -53.868134   
2                         NaN                         NaN   
3                         NaN                         NaN   
4                         NaN                         NaN   

   L3_CH4_solar_zenith_angle  
0                  22.545118  
1                  19.293652  
2                        NaN  
3                        NaN  
4                        NaN  

[5 rows x 82 columns]

Shape-Training Dataset
df_train (30557, 82)
Data Types :
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30557 entries, 0 to 30556
Data columns (total 82 columns):
 #   Column                                               Non-Null Count  Dtype  
---  ------                                               --------------  -----  
 0   Place_ID X Date                                      30557 non-null  object 
 1   Date                                                 30557 non-null  object 
 2   Place_ID                                             30557 non-null  object 
 3   target                                               30557 non-null  float64
 4   target_min                                           30557 non-null  float64
 5   target_max                                           30557 non-null  float64
 6   target_variance                                      30557 non-null  float64
 7   target_count                                         30557 non-null  int64  
 8   precipita

In [4]:
df_test.head()
df_test.shape

(16136, 77)

In [5]:
df_sub.head()

Place_ID X Date  target
0  0OS9LVX X 2020-01-02       0
1  0OS9LVX X 2020-01-03       0
2  0OS9LVX X 2020-01-04       0
3  0OS9LVX X 2020-01-05       0
4  0OS9LVX X 2020-01-06       0

In [6]:
# prompt: check null or missing values in df_train ,show only where there is missing or null

print('Null values in training dataset')
print(df_train.isnull().sum()[df_train.isnull().sum() > 0])
print("----------------------------------------------------")
print('Null values in test dataset')
print(df_test.isnull().sum()[df_test.isnull().sum() > 0])

Null values in training dataset
L3_NO2_NO2_column_number_density           2189
L3_NO2_NO2_slant_column_number_density     2189
L3_NO2_absorbing_aerosol_index             2189
L3_NO2_cloud_fraction                      2189
L3_NO2_sensor_altitude                     2189
                                          ...  
L3_CH4_aerosol_optical_depth              24765
L3_CH4_sensor_azimuth_angle               24765
L3_CH4_sensor_zenith_angle                24765
L3_CH4_solar_azimuth_angle                24765
L3_CH4_solar_zenith_angle                 24765
Length: 68, dtype: int64
----------------------------------------------------
Null values in test dataset
L3_NO2_NO2_column_number_density           1327
L3_NO2_NO2_slant_column_number_density     1327
L3_NO2_absorbing_aerosol_index             1327
L3_NO2_cloud_fraction                      1327
L3_NO2_sensor_altitude                     1327
                                          ...  
L3_CH4_aerosol_optical_depth              1302

In [7]:
# fill the missing values

for col in df_train.columns:
  if df_train[col].dtype == 'float64' and df_train[col].isnull().any():
    # Option 1: Fill with the mean of the column
    #df_train[col].fillna(df_train[col].mean(), inplace=True)
    # Option 2: Fill with the median of the column (often more robust to outliers)
    df_train[col].fillna(df_train[col].median(), inplace=True)


# Repeat for the test set
for col in df_test.columns:
  if df_test[col].dtype == 'float64' and df_test[col].isnull().any():

    #df_test[col].fillna(df_train[col].mean(), inplace=True)
    df_test[col].fillna(df_train[col].median(), inplace=True)



print('\nNull values after filling in training dataset')
print(df_train.isnull().sum()[df_train.isnull().sum() > 0])

print('\nNull values after filling in test dataset')
print(df_test.isnull().sum()[df_test.isnull().sum() > 0])



Null values after filling in training dataset
Series([], dtype: int64)

Null values after filling in test dataset
Series([], dtype: int64)


<ipython-input-7-7b49da92423c>:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_train[col].fillna(df_train[col].median(), inplace=True)
<ipython-input-7-7b49da92423c>:16: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)',

In [8]:
# Transform test dataset

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np # Import numpy to use sqrt

# Define features (X) and target (y)
# Assuming 'target' is the name of your target variable
# Exclude 'ID' or any other non-feature columns from features
# Exclude 'Place_ID X Date' as it is a string column
features = [col for col in df_train.columns if col not in ['target','Place_ID X Date',"Date","Place_ID"]]
X = df_train[features]
y = df_train['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) # 80% train, 20% test

# Initialize and train the Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse}")
print(f"Root Mean Squared Error (RMSE): {rmse}")
print(f"R-squared (R2): {r2}")

# Calculate additional statistics on the target variable in the training set
print("\nTarget variable statistics:")
print(f"Target minimum: {y_train.min()}")
print(f"Target maximum: {y_train.max()}")
print(f"Target variance: {y_train.var()}")
print(f"Target count: {y_train.count()}")

Mean Squared Error (MSE): 439.5793888646482
Root Mean Squared Error (RMSE): 20.96614864167113
R-squared (R2): 0.8002005637562787

Target variable statistics:
Target minimum: 1.0
Target maximum: 815.0
Target variance: 2194.943463739312
Target count: 24445


### Data with missing values as median
Mean Squared Error (MSE): 439.5793888646482
Root Mean Squared Error (RMSE): 20.96614864167113
R-squared (R2): 0.8002005637562787

Target variable statistics:
Target minimum: 1.0
Target maximum: 815.0
Target variance: 2194.943463739312
Target count: 24445

In [10]:
# prompt: create a csv file with test dataset similar to df_sub dataset

# Create a synthetic df_sub-like DataFrame
#data = {'Place_ID X Date': [f'place_{i} X 2023-01-{j:02d}' for i in range(10) for j in range(1, 31)],
        #'target': [np.random.rand() for _ in range(10 * 30)]} # Example target values

#df_sub_test = pd.DataFrame(data)

# Save the DataFrame to a CSV file
#csv_file_path = '/content/drive/My Drive/Colab Notebooks/DF_Zindi_Urban-Air/SampleSubmission_Test.csv'
#df_sub_test.to_csv(csv_file_path, index=False)

#print(f"Created synthetic df_sub test file at: {csv_file_path}")
#print(df_sub_test.head())
